src/VisualStudio/Core/Def/Implementation/Remote/VisualStudioWorkspaceServiceHubConnector.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+
+namespace Microsoft.VisualStudio.LanguageServices.Remote
+{
+    /// <summary>
+    /// Connects <see cref="VisualStudioWorkspace"/> to the ServiceHub services.
+    /// Launches ServiceHub if it is not running yet and starts services that push information from <see cref="VisualStudioWorkspace"/> to the ServiceHub process.
+    /// </summary>
+    [ExportEventListener(WellKnownEventListeners.Workspace, WorkspaceKind.Host), Shared]
+    internal sealed class VisualStudioWorkspaceServiceHubConnector : IEventListener<object>, IEventListenerStoppable
```

Comment:
interesting.  i didn't know about this pattern.  should i use it with the OOP services i created that are currently BG loaded in the Package?  When do these events fire?

Reply:
I think it would be good to use it. You can probably just move them to this type. The StartListening event fires on first workspace event (e.g. SolutionAdded):

```
	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Remote.VisualStudioWorkspaceServiceHubConnector.StartListening(Microsoft.CodeAnalysis.Workspace workspace, object serviceOpt) Line 43	C#
>	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Host.DefaultWorkspaceEventListenerServiceFactory.Service.EnsureListeners() Line 74	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.GetEventHandlers<Microsoft.CodeAnalysis.WorkspaceChangeEventArgs>(string eventName) Line 214	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.RaiseWorkspaceChangedEventAsync(Microsoft.CodeAnalysis.WorkspaceChangeKind kind, Microsoft.CodeAnalysis.Solution oldSolution, Microsoft.CodeAnalysis.Solution newSolution, Microsoft.CodeAnalysis.ProjectId projectId, Microsoft.CodeAnalysis.DocumentId documentId) Line 58	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.OnSolutionAdded(Microsoft.CodeAnalysis.SolutionInfo solutionInfo) Line 397	C#
```

`EnsureListeners` makes sure it only sends the event once per workspace.

Reply:
> You can probably just move them to this type

Sure, if you don't mind that, i'll do that after you merge this in :)